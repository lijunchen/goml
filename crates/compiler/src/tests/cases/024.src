struct Point {
  x: int,
  y: int
}

fn point_to_string(p: Point) -> string {
  let Point {x : x, y: y} = p in
  string_add("Point { x: ", string_add(int_to_string(x), string_add(", y: ", string_add(int_to_string(y), " }"))))
}

enum Color {
  Red,
  Green,
  Blue,
}

fn color_to_string(c: Color) -> string {
  match c {
    Red => "Red",
    Green => "Green",
    Blue => "Blue",
  }
}

struct Line {
  from: Point,
  to: Point,
  color: Color,
}

fn line_to_string(l: Line) -> string {
  let Line { from: from, to: to, color: color } = l in
  string_add("Line { from: ", string_add(point_to_string(from), string_add(", to: ", string_add(point_to_string(to), string_add(", color: ", string_add(color_to_string(color), " }"))))))
}

fn main() {
  let p0 = Point { x: 0, y: 0 } in
  let p1 = Point { x: 10, y: 10 } in
  let line = Line { from: p0, to: p1, color: Red } in
  string_println(line_to_string(line))
}
