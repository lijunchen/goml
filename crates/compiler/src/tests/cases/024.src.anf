fn point_to_string(p/0: Point) -> string {
  let mtmp0 = p/0 in
  let x1 = Point.x(mtmp0) in
  let x2 = Point.y(mtmp0) in
  let y/2 = x2 in
  let x/1 = x1 in
  let t8 = int_to_string(x/1) in
  let t11 = int_to_string(y/2) in
  let t10 = string_add(t11, " }") in
  let t9 = string_add(", y: ", t10) in
  let t7 = string_add(t8, t9) in
  string_add("Point { x: ", t7)
}

fn color_to_string(c/3: Color) -> string {
  match c/3 {
    Tag_0 => {
      "Red"
    },
    Tag_1 => {
      "Green"
    },
    Tag_2 => {
      "Blue"
    },
  }
}

fn line_to_string(l/4: Line) -> string {
  let mtmp3 = l/4 in
  let x4 = Line.from(mtmp3) in
  let x5 = Line.to(mtmp3) in
  let x6 = Line.color(mtmp3) in
  let color/7 = x6 in
  let to/6 = x5 in
  let from/5 = x4 in
  let t13 = point_to_string(from/5) in
  let t16 = point_to_string(to/6) in
  let t19 = color_to_string(color/7) in
  let t18 = string_add(t19, " }") in
  let t17 = string_add(", color: ", t18) in
  let t15 = string_add(t16, t17) in
  let t14 = string_add(", to: ", t15) in
  let t12 = string_add(t13, t14) in
  string_add("Line { from: ", t12)
}

fn main() -> unit {
  let p0/8 = Point { x: 0, y: 0 } in
  let p1/9 = Point { x: 10, y: 10 } in
  let t20 = Tag_0 in
  let line/10 = Line { from: p0/8, to: p1/9, color: t20 } in
  let t21 = line_to_string(line/10) in
  string_println(t21)
}