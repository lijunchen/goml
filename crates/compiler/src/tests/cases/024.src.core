fn make_point() -> Point {
  Point(0, 0)
}

fn flip(point/0: Point) -> Point {
  let mtmp0 = point/0 in
  let x1 = Point.x(mtmp0) in
  let x2 = Point.y(mtmp0) in
  let y/2 = x2 in
  let x/1 = x1 in
  Point(y/2, x/1)
}

fn wrap_int(x/3: int) -> Wrapper[int] {
  Wrapper(x/3)
}

fn main() -> unit {
  let start/4 = make_point() in
  let swapped/5 = flip(Point(1, 2)) in
  let boxed/6 = wrap_int(3) in
  let point_box/7 = Wrapper(swapped/5) in
  let mtmp3 = start/4 in
  let mtmp4 = boxed/6 in
  let mtmp5 = point_box/7 in
  ()
}