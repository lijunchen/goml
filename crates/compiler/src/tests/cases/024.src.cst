FILE@0..913
  STRUCT@0..37
    StructKeyword@0..6 "struct"
    Whitespace@6..7 " "
    Uident@7..12 "Point"
    Whitespace@12..13 " "
    LBrace@13..14 "{"
    Whitespace@14..17 "\n  "
    STRUCT_FIELD_LIST@17..37
      STRUCT_FIELD@17..23
        Lident@17..18 "x"
        Colon@18..19 ":"
        Whitespace@19..20 " "
        TYPE_INT@20..23
          IntKeyword@20..23 "int"
      Comma@23..24 ","
      Whitespace@24..27 "\n  "
      STRUCT_FIELD@27..34
        Lident@27..28 "y"
        Colon@28..29 ":"
        Whitespace@29..30 " "
        TYPE_INT@30..34
          IntKeyword@30..33 "int"
          Whitespace@33..34 "\n"
      RBrace@34..35 "}"
      Whitespace@35..37 "\n\n"
  FN@37..229
    FnKeyword@37..39 "fn"
    Whitespace@39..40 " "
    Lident@40..55 "point_to_string"
    PARAM_LIST@55..66
      LParen@55..56 "("
      PARAM@56..64
        Lident@56..57 "p"
        Colon@57..58 ":"
        Whitespace@58..59 " "
        TYPE_TAPP@59..64
          Uident@59..64 "Point"
      RParen@64..65 ")"
      Whitespace@65..66 " "
    Arrow@66..68 "->"
    Whitespace@68..69 " "
    TYPE_STRING@69..76
      StringKeyword@69..75 "string"
      Whitespace@75..76 " "
    BLOCK@76..229
      LBrace@76..77 "{"
      Whitespace@77..80 "\n  "
      EXPR_LET@80..226
        LetKeyword@80..83 "let"
        Whitespace@83..84 " "
        PATTERN_CONSTR@84..104
          Uident@84..89 "Point"
          Whitespace@89..90 " "
          STRUCT_PATTERN_FIELD_LIST@90..104
            LBrace@90..91 "{"
            STRUCT_PATTERN_FIELD@91..96
              Lident@91..92 "x"
              Whitespace@92..93 " "
              Colon@93..94 ":"
              Whitespace@94..95 " "
              PATTERN_VARIABLE@95..96
                Lident@95..96 "x"
            Comma@96..97 ","
            Whitespace@97..98 " "
            STRUCT_PATTERN_FIELD@98..102
              Lident@98..99 "y"
              Colon@99..100 ":"
              Whitespace@100..101 " "
              PATTERN_VARIABLE@101..102
                Lident@101..102 "y"
            RBrace@102..103 "}"
            Whitespace@103..104 " "
        Eq@104..105 "="
        Whitespace@105..106 " "
        EXPR_LET_VALUE@106..108
          EXPR_LIDENT@106..108
            Lident@106..107 "p"
            Whitespace@107..108 " "
        InKeyword@108..110 "in"
        Whitespace@110..113 "\n  "
        EXPR_LET_BODY@113..226
          EXPR_CALL@113..226
            EXPR_LIDENT@113..123
              Lident@113..123 "string_add"
            ARG_LIST@123..226
              LParen@123..124 "("
              ARG@124..139
                EXPR_STR@124..137
                  Str@124..137 "\"Point { x: \""
                Comma@137..138 ","
                Whitespace@138..139 " "
              ARG@139..224
                EXPR_CALL@139..224
                  EXPR_LIDENT@139..149
                    Lident@139..149 "string_add"
                  ARG_LIST@149..224
                    LParen@149..150 "("
                    ARG@150..168
                      EXPR_CALL@150..166
                        EXPR_LIDENT@150..163
                          Lident@150..163 "int_to_string"
                        ARG_LIST@163..166
                          LParen@163..164 "("
                          ARG@164..165
                            EXPR_LIDENT@164..165
                              Lident@164..165 "x"
                          RParen@165..166 ")"
                      Comma@166..167 ","
                      Whitespace@167..168 " "
                    ARG@168..223
                      EXPR_CALL@168..223
                        EXPR_LIDENT@168..178
                          Lident@168..178 "string_add"
                        ARG_LIST@178..223
                          LParen@178..179 "("
                          ARG@179..188
                            EXPR_STR@179..186
                              Str@179..186 "\", y: \""
                            Comma@186..187 ","
                            Whitespace@187..188 " "
                          ARG@188..222
                            EXPR_CALL@188..222
                              EXPR_LIDENT@188..198
                                Lident@188..198 "string_add"
                              ARG_LIST@198..222
                                LParen@198..199 "("
                                ARG@199..217
                                  EXPR_CALL@199..215
                                    EXPR_LIDENT@199..212
                                      Lident@199..212 "int_to_string"
                                    ARG_LIST@212..215
                                      LParen@212..213 "("
                                      ARG@213..214
                                        EXPR_LIDENT@213..214
                                          Lident@213..214 "y"
                                      RParen@214..215 ")"
                                  Comma@215..216 ","
                                  Whitespace@216..217 " "
                                ARG@217..221
                                  EXPR_STR@217..221
                                    Str@217..221 "\" }\""
                                RParen@221..222 ")"
                          RParen@222..223 ")"
                    RParen@223..224 ")"
              RParen@224..225 ")"
              Whitespace@225..226 "\n"
      RBrace@226..227 "}"
      Whitespace@227..229 "\n\n"
  ENUM@229..269
    EnumKeyword@229..233 "enum"
    Whitespace@233..234 " "
    Uident@234..239 "Color"
    Whitespace@239..240 " "
    LBrace@240..241 "{"
    Whitespace@241..244 "\n  "
    VARIANT_LIST@244..269
      VARIANT@244..247
        Uident@244..247 "Red"
      Comma@247..248 ","
      Whitespace@248..251 "\n  "
      VARIANT@251..256
        Uident@251..256 "Green"
      Comma@256..257 ","
      Whitespace@257..260 "\n  "
      VARIANT@260..264
        Uident@260..264 "Blue"
      Comma@264..265 ","
      Whitespace@265..266 "\n"
      RBrace@266..267 "}"
      Whitespace@267..269 "\n\n"
  FN@269..389
    FnKeyword@269..271 "fn"
    Whitespace@271..272 " "
    Lident@272..287 "color_to_string"
    PARAM_LIST@287..298
      LParen@287..288 "("
      PARAM@288..296
        Lident@288..289 "c"
        Colon@289..290 ":"
        Whitespace@290..291 " "
        TYPE_TAPP@291..296
          Uident@291..296 "Color"
      RParen@296..297 ")"
      Whitespace@297..298 " "
    Arrow@298..300 "->"
    Whitespace@300..301 " "
    TYPE_STRING@301..308
      StringKeyword@301..307 "string"
      Whitespace@307..308 " "
    BLOCK@308..389
      LBrace@308..309 "{"
      Whitespace@309..312 "\n  "
      EXPR_MATCH@312..386
        MatchKeyword@312..317 "match"
        Whitespace@317..318 " "
        EXPR_LIDENT@318..320
          Lident@318..319 "c"
          Whitespace@319..320 " "
        MATCH_ARM_LIST@320..386
          LBrace@320..321 "{"
          Whitespace@321..326 "\n    "
          MATCH_ARM@326..338
            PATTERN_CONSTR@326..330
              Uident@326..329 "Red"
              Whitespace@329..330 " "
            FatArrow@330..332 "=>"
            Whitespace@332..333 " "
            EXPR_STR@333..338
              Str@333..338 "\"Red\""
          Comma@338..339 ","
          Whitespace@339..344 "\n    "
          MATCH_ARM@344..360
            PATTERN_CONSTR@344..350
              Uident@344..349 "Green"
              Whitespace@349..350 " "
            FatArrow@350..352 "=>"
            Whitespace@352..353 " "
            EXPR_STR@353..360
              Str@353..360 "\"Green\""
          Comma@360..361 ","
          Whitespace@361..366 "\n    "
          MATCH_ARM@366..380
            PATTERN_CONSTR@366..371
              Uident@366..370 "Blue"
              Whitespace@370..371 " "
            FatArrow@371..373 "=>"
            Whitespace@373..374 " "
            EXPR_STR@374..380
              Str@374..380 "\"Blue\""
          Comma@380..381 ","
          Whitespace@381..384 "\n  "
          RBrace@384..385 "}"
          Whitespace@385..386 "\n"
      RBrace@386..387 "}"
      Whitespace@387..389 "\n\n"
  STRUCT@389..450
    StructKeyword@389..395 "struct"
    Whitespace@395..396 " "
    Uident@396..400 "Line"
    Whitespace@400..401 " "
    LBrace@401..402 "{"
    Whitespace@402..405 "\n  "
    STRUCT_FIELD_LIST@405..450
      STRUCT_FIELD@405..416
        Lident@405..409 "from"
        Colon@409..410 ":"
        Whitespace@410..411 " "
        TYPE_TAPP@411..416
          Uident@411..416 "Point"
      Comma@416..417 ","
      Whitespace@417..420 "\n  "
      STRUCT_FIELD@420..429
        Lident@420..422 "to"
        Colon@422..423 ":"
        Whitespace@423..424 " "
        TYPE_TAPP@424..429
          Uident@424..429 "Point"
      Comma@429..430 ","
      Whitespace@430..433 "\n  "
      STRUCT_FIELD@433..445
        Lident@433..438 "color"
        Colon@438..439 ":"
        Whitespace@439..440 " "
        TYPE_TAPP@440..445
          Uident@440..445 "Color"
      Comma@445..446 ","
      Whitespace@446..447 "\n"
      RBrace@447..448 "}"
      Whitespace@448..450 "\n\n"
  FN@450..734
    FnKeyword@450..452 "fn"
    Whitespace@452..453 " "
    Lident@453..467 "line_to_string"
    PARAM_LIST@467..477
      LParen@467..468 "("
      PARAM@468..475
        Lident@468..469 "l"
        Colon@469..470 ":"
        Whitespace@470..471 " "
        TYPE_TAPP@471..475
          Uident@471..475 "Line"
      RParen@475..476 ")"
      Whitespace@476..477 " "
    Arrow@477..479 "->"
    Whitespace@479..480 " "
    TYPE_STRING@480..487
      StringKeyword@480..486 "string"
      Whitespace@486..487 " "
    BLOCK@487..734
      LBrace@487..488 "{"
      Whitespace@488..491 "\n  "
      EXPR_LET@491..731
        LetKeyword@491..494 "let"
        Whitespace@494..495 " "
        PATTERN_CONSTR@495..537
          Uident@495..499 "Line"
          Whitespace@499..500 " "
          STRUCT_PATTERN_FIELD_LIST@500..537
            LBrace@500..501 "{"
            Whitespace@501..502 " "
            STRUCT_PATTERN_FIELD@502..512
              Lident@502..506 "from"
              Colon@506..507 ":"
              Whitespace@507..508 " "
              PATTERN_VARIABLE@508..512
                Lident@508..512 "from"
            Comma@512..513 ","
            Whitespace@513..514 " "
            STRUCT_PATTERN_FIELD@514..520
              Lident@514..516 "to"
              Colon@516..517 ":"
              Whitespace@517..518 " "
              PATTERN_VARIABLE@518..520
                Lident@518..520 "to"
            Comma@520..521 ","
            Whitespace@521..522 " "
            STRUCT_PATTERN_FIELD@522..535
              Lident@522..527 "color"
              Colon@527..528 ":"
              Whitespace@528..529 " "
              PATTERN_VARIABLE@529..535
                Lident@529..534 "color"
                Whitespace@534..535 " "
            RBrace@535..536 "}"
            Whitespace@536..537 " "
        Eq@537..538 "="
        Whitespace@538..539 " "
        EXPR_LET_VALUE@539..541
          EXPR_LIDENT@539..541
            Lident@539..540 "l"
            Whitespace@540..541 " "
        InKeyword@541..543 "in"
        Whitespace@543..546 "\n  "
        EXPR_LET_BODY@546..731
          EXPR_CALL@546..731
            EXPR_LIDENT@546..556
              Lident@546..556 "string_add"
            ARG_LIST@556..731
              LParen@556..557 "("
              ARG@557..574
                EXPR_STR@557..572
                  Str@557..572 "\"Line { from: \""
                Comma@572..573 ","
                Whitespace@573..574 " "
              ARG@574..729
                EXPR_CALL@574..729
                  EXPR_LIDENT@574..584
                    Lident@574..584 "string_add"
                  ARG_LIST@584..729
                    LParen@584..585 "("
                    ARG@585..608
                      EXPR_CALL@585..606
                        EXPR_LIDENT@585..600
                          Lident@585..600 "point_to_string"
                        ARG_LIST@600..606
                          LParen@600..601 "("
                          ARG@601..605
                            EXPR_LIDENT@601..605
                              Lident@601..605 "from"
                          RParen@605..606 ")"
                      Comma@606..607 ","
                      Whitespace@607..608 " "
                    ARG@608..728
                      EXPR_CALL@608..728
                        EXPR_LIDENT@608..618
                          Lident@608..618 "string_add"
                        ARG_LIST@618..728
                          LParen@618..619 "("
                          ARG@619..629
                            EXPR_STR@619..627
                              Str@619..627 "\", to: \""
                            Comma@627..628 ","
                            Whitespace@628..629 " "
                          ARG@629..727
                            EXPR_CALL@629..727
                              EXPR_LIDENT@629..639
                                Lident@629..639 "string_add"
                              ARG_LIST@639..727
                                LParen@639..640 "("
                                ARG@640..661
                                  EXPR_CALL@640..659
                                    EXPR_LIDENT@640..655
                                      Lident@640..655 "point_to_string"
                                    ARG_LIST@655..659
                                      LParen@655..656 "("
                                      ARG@656..658
                                        EXPR_LIDENT@656..658
                                          Lident@656..658 "to"
                                      RParen@658..659 ")"
                                  Comma@659..660 ","
                                  Whitespace@660..661 " "
                                ARG@661..726
                                  EXPR_CALL@661..726
                                    EXPR_LIDENT@661..671
                                      Lident@661..671 "string_add"
                                    ARG_LIST@671..726
                                      LParen@671..672 "("
                                      ARG@672..685
                                        EXPR_STR@672..683
                                          Str@672..683 "\", color: \""
                                        Comma@683..684 ","
                                        Whitespace@684..685 " "
                                      ARG@685..725
                                        EXPR_CALL@685..725
                                          EXPR_LIDENT@685..695
                                            Lident@685..695 "string_add"
                                          ARG_LIST@695..725
                                            LParen@695..696 "("
                                            ARG@696..720
                                              EXPR_CALL@696..718
                                                EXPR_LIDENT@696..711
                                                  Lident@696..711 "color_to_string"
                                                ARG_LIST@711..718
                                                  LParen@711..712 "("
                                                  ARG@712..717
                                                    EXPR_LIDENT@712..717
                                                      Lident@712..717 "color"
                                                  RParen@717..718 ")"
                                              Comma@718..719 ","
                                              Whitespace@719..720 " "
                                            ARG@720..724
                                              EXPR_STR@720..724
                                                Str@720..724 "\" }\""
                                            RParen@724..725 ")"
                                      RParen@725..726 ")"
                                RParen@726..727 ")"
                          RParen@727..728 ")"
                    RParen@728..729 ")"
              RParen@729..730 ")"
              Whitespace@730..731 "\n"
      RBrace@731..732 "}"
      Whitespace@732..734 "\n\n"
  FN@734..913
    FnKeyword@734..736 "fn"
    Whitespace@736..737 " "
    Lident@737..741 "main"
    PARAM_LIST@741..744
      LParen@741..742 "("
      RParen@742..743 ")"
      Whitespace@743..744 " "
    BLOCK@744..913
      LBrace@744..745 "{"
      Whitespace@745..748 "\n  "
      EXPR_LET@748..911
        LetKeyword@748..751 "let"
        Whitespace@751..752 " "
        PATTERN_VARIABLE@752..755
          Lident@752..754 "p0"
          Whitespace@754..755 " "
        Eq@755..756 "="
        Whitespace@756..757 " "
        EXPR_LET_VALUE@757..778
          EXPR_STRUCT_LITERAL@757..778
            Uident@757..762 "Point"
            Whitespace@762..763 " "
            STRUCT_LITERAL_FIELD_LIST@763..778
              LBrace@763..764 "{"
              Whitespace@764..765 " "
              STRUCT_LITERAL_FIELD@765..769
                Lident@765..766 "x"
                Colon@766..767 ":"
                Whitespace@767..768 " "
                EXPR_INT@768..769
                  Int@768..769 "0"
              Comma@769..770 ","
              Whitespace@770..771 " "
              STRUCT_LITERAL_FIELD@771..776
                Lident@771..772 "y"
                Colon@772..773 ":"
                Whitespace@773..774 " "
                EXPR_INT@774..776
                  Int@774..775 "0"
                  Whitespace@775..776 " "
              RBrace@776..777 "}"
              Whitespace@777..778 " "
        InKeyword@778..780 "in"
        Whitespace@780..783 "\n  "
        EXPR_LET_BODY@783..911
          EXPR_LET@783..911
            LetKeyword@783..786 "let"
            Whitespace@786..787 " "
            PATTERN_VARIABLE@787..790
              Lident@787..789 "p1"
              Whitespace@789..790 " "
            Eq@790..791 "="
            Whitespace@791..792 " "
            EXPR_LET_VALUE@792..815
              EXPR_STRUCT_LITERAL@792..815
                Uident@792..797 "Point"
                Whitespace@797..798 " "
                STRUCT_LITERAL_FIELD_LIST@798..815
                  LBrace@798..799 "{"
                  Whitespace@799..800 " "
                  STRUCT_LITERAL_FIELD@800..805
                    Lident@800..801 "x"
                    Colon@801..802 ":"
                    Whitespace@802..803 " "
                    EXPR_INT@803..805
                      Int@803..805 "10"
                  Comma@805..806 ","
                  Whitespace@806..807 " "
                  STRUCT_LITERAL_FIELD@807..813
                    Lident@807..808 "y"
                    Colon@808..809 ":"
                    Whitespace@809..810 " "
                    EXPR_INT@810..813
                      Int@810..812 "10"
                      Whitespace@812..813 " "
                  RBrace@813..814 "}"
                  Whitespace@814..815 " "
            InKeyword@815..817 "in"
            Whitespace@817..820 "\n  "
            EXPR_LET_BODY@820..911
              EXPR_LET@820..911
                LetKeyword@820..823 "let"
                Whitespace@823..824 " "
                PATTERN_VARIABLE@824..829
                  Lident@824..828 "line"
                  Whitespace@828..829 " "
                Eq@829..830 "="
                Whitespace@830..831 " "
                EXPR_LET_VALUE@831..869
                  EXPR_STRUCT_LITERAL@831..869
                    Uident@831..835 "Line"
                    Whitespace@835..836 " "
                    STRUCT_LITERAL_FIELD_LIST@836..869
                      LBrace@836..837 "{"
                      Whitespace@837..838 " "
                      STRUCT_LITERAL_FIELD@838..846
                        Lident@838..842 "from"
                        Colon@842..843 ":"
                        Whitespace@843..844 " "
                        EXPR_LIDENT@844..846
                          Lident@844..846 "p0"
                      Comma@846..847 ","
                      Whitespace@847..848 " "
                      STRUCT_LITERAL_FIELD@848..854
                        Lident@848..850 "to"
                        Colon@850..851 ":"
                        Whitespace@851..852 " "
                        EXPR_LIDENT@852..854
                          Lident@852..854 "p1"
                      Comma@854..855 ","
                      Whitespace@855..856 " "
                      STRUCT_LITERAL_FIELD@856..867
                        Lident@856..861 "color"
                        Colon@861..862 ":"
                        Whitespace@862..863 " "
                        EXPR_UIDENT@863..867
                          Uident@863..866 "Red"
                          Whitespace@866..867 " "
                      RBrace@867..868 "}"
                      Whitespace@868..869 " "
                InKeyword@869..871 "in"
                Whitespace@871..874 "\n  "
                EXPR_LET_BODY@874..911
                  EXPR_CALL@874..911
                    EXPR_LIDENT@874..888
                      Lident@874..888 "string_println"
                    ARG_LIST@888..911
                      LParen@888..889 "("
                      ARG@889..909
                        EXPR_CALL@889..909
                          EXPR_LIDENT@889..903
                            Lident@889..903 "line_to_string"
                          ARG_LIST@903..909
                            LParen@903..904 "("
                            ARG@904..908
                              EXPR_LIDENT@904..908
                                Lident@904..908 "line"
                            RParen@908..909 ")"
                      RParen@909..910 ")"
                      Whitespace@910..911 "\n"
      RBrace@911..912 "}"
      Whitespace@912..913 "\n"