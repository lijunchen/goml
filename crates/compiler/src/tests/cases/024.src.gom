package main

import (
    "fmt"
)

func unit_to_string(x struct{}) string {
    return "()"
}

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_sub(x int, y int) int {
    return x - y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_add(a string, b string) string {
    return a + b
}

func string_print(s string) struct{} {
    fmt.Print(s)
    return struct{}{}
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func missing(s string) struct{} {
    panic("missing: " + s)
    return struct{}{}
}

type Point struct {
    x int
    y int
}

type Line struct {
    from Point
    to Point
    color Color
}

type Color interface {
    isColor()
}

type Red struct {}

func (_ Red) isColor() {}

type Green struct {}

func (_ Green) isColor() {}

type Blue struct {}

func (_ Blue) isColor() {}

func point_to_string(p__0 Point) string {
    var ret22 string
    var mtmp0 Point = p__0
    var x1 int = mtmp0.x
    var x2 int = mtmp0.y
    var y__2 int = x2
    var x__1 int = x1
    var t8 string = int_to_string(x__1)
    var t11 string = int_to_string(y__2)
    var t10 string = string_add(t11, " }")
    var t9 string = string_add(", y: ", t10)
    var t7 string = string_add(t8, t9)
    ret22 = string_add("Point { x: ", t7)
    return ret22
}

func color_to_string(c__3 Color) string {
    var ret23 string
    switch c__3.(type) {
    case Red:
        ret23 = "Red"
    case Green:
        ret23 = "Green"
    case Blue:
        ret23 = "Blue"
    }
    return ret23
}

func line_to_string(l__4 Line) string {
    var ret24 string
    var mtmp3 Line = l__4
    var x4 Point = mtmp3.from
    var x5 Point = mtmp3.to
    var x6 Color = mtmp3.color
    var color__7 Color = x6
    var to__6 Point = x5
    var from__5 Point = x4
    var t13 string = point_to_string(from__5)
    var t16 string = point_to_string(to__6)
    var t19 string = color_to_string(color__7)
    var t18 string = string_add(t19, " }")
    var t17 string = string_add(", color: ", t18)
    var t15 string = string_add(t16, t17)
    var t14 string = string_add(", to: ", t15)
    var t12 string = string_add(t13, t14)
    ret24 = string_add("Line { from: ", t12)
    return ret24
}

func main0() struct{} {
    var ret25 struct{}
    var p0__8 Point = Point{
        x: 0,
        y: 0,
    }
    var p1__9 Point = Point{
        x: 10,
        y: 10,
    }
    var t20 Color = Red{}
    var line__10 Line = Line{
        from: p0__8,
        to: p1__9,
        color: t20,
    }
    var t21 string = line_to_string(line__10)
    ret25 = string_println(t21)
    return ret25
}

func main() {
    main0()
}
