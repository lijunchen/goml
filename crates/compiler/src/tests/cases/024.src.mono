fn point_to_string(p/0: Point) -> string {
  let mtmp0 = p/0 in
  let x1 = Point.x(mtmp0) in
  let x2 = Point.y(mtmp0) in
  let y/2 = x2 in
  let x/1 = x1 in
  string_add("Point { x: ", string_add(int_to_string(x/1), string_add(", y: ", string_add(int_to_string(y/2), " }"))))
}

fn color_to_string(c/3: Color) -> string {
  match c/3 {
    Color::Red => {
      "Red"
    },
    Color::Green => {
      "Green"
    },
    Color::Blue => {
      "Blue"
    },
  }
}

fn line_to_string(l/4: Line) -> string {
  let mtmp3 = l/4 in
  let x4 = Line.from(mtmp3) in
  let x5 = Line.to(mtmp3) in
  let x6 = Line.color(mtmp3) in
  let color/7 = x6 in
  let to/6 = x5 in
  let from/5 = x4 in
  string_add("Line { from: ", string_add(point_to_string(from/5), string_add(", to: ", string_add(point_to_string(to/6), string_add(", color: ", string_add(color_to_string(color/7), " }"))))))
}

fn main() -> unit {
  let p0/8 = Point { x: 0, y: 0 } in
  let p1/9 = Point { x: 10, y: 10 } in
  let line/10 = Line { from: p0/8, to: p1/9, color: Color::Red } in
  string_println(line_to_string(line/10))
}