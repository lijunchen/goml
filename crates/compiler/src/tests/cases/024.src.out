Line { from: Point { x: 0, y: 0 }, to: Point { x: 10, y: 10 }, color: Red }
